heat/core/statistics.py

```diff
@@ -626,77 +627,73 @@ def maximum(x1, x2, out=None):
-
-    # apply split semantics
-    if x1.split is not None or x2.split is not None:
-        if x1.split is None:
-            x1.resplit_(x2.split)
-        if x2.split is None:
-            x2.resplit_(x1.split)
-        if x1.split != x2.split:
-            if np.prod(x1.gshape) < np.prod(x2.gshape):
-                x1.resplit_(x2.split)
-            if np.prod(x2.gshape) < np.prod(x1.gshape):
-                x2.resplit_(x1.split)
-            else:
-                if x1.split < x2.split:
-                    x2.resplit_(x1.split)
-                else:
-                    x1.resplit_(x2.split)
-        split = x1.split
-    else:
-        split = None
 
-    # locally: apply torch.max(x1, x2)
-    output_lshape = stride_tricks.broadcast_shape(x1.lshape, x2.lshape)
-    lresult = factories.empty(output_lshape, dtype=x1.dtype)
-    lresult._DNDarray__array = torch.max(x1._DNDarray__array, x2._DNDarray__array)
-    lresult._DNDarray__dtype = types.promote_types(x1.dtype, x2.dtype)
-    lresult._DNDarray__split = split
-    if x1.split is not None or x2.split is not None:
-        if x1.comm.is_distributed():  # assuming x1.comm = x2.comm
-            output_gshape = stride_tricks.broadcast_shape(x1.gshape, x2.gshape)
-            result = factories.empty(output_gshape, dtype=x1.dtype)
-            x1.comm.Allgather(lresult, result)
-            # TODO: adopt Allgatherv() as soon as it is fixed, Issue #233
-            result._DNDarray__dtype = lresult._DNDarray__dtype
-            result._DNDarray__split = split
-
-            if out is not None:
-                if out.shape != output_gshape:
-                    raise ValueError(
-                        "Expecting output buffer of shape {}, got {}".format(
-                            output_gshape, out.shape
-                        )
-                    )
-                out._DNDarray__array = result._DNDarray__array
-                out._DNDarray__dtype = result._DNDarray__dtype
-                out._DNDarray__split = split
-                out._DNDarray__device = x1.device
-                out._DNDarray__comm = x1.comm
+    # sanitize split axes and shapes
+    split1, size1, shape1, comm1 = x1.split, x1.size, x1.gshape, x1.comm
+    split2, size2, shape2, comm2 = x2.split, x2.size, x2.gshape, x2.comm
+    output_gshape = stride_tricks.broadcast_shape(shape1, shape2)
 
-                return out
-            return result
+    # DNDarrays must be distributed along same axis
+    if split1 is not None and split2 is not None and split1 != split2:
+        raise ValueError(
+            "x1 and x2 must be distributed along the same axis, but x1.split is {}, x2.split is {}".format(
+                split1, split2
+            )
+        )
+    # Exception: split axes can differ if one of the DNDarrays is a scalar
+    if split1 is not None or split2 is not None:
+        if split1 is None and size1 > 1 or split2 is None and size2 > 1:
```

Comment:
What happens if one DNDarray is empty? Admittedly, this probably wouldn't make a huge difference for splits, but for pedantic reasons I would replace the > with !=

Reply:
hmm, if the DNDarray is empty, the function should raise an error much earlier. Thanks for the reminder.

Otherwise, `size` is the total number of elements of the DNDarray so it can't be less than 1.

Reply:
Alright, then I take my back my criticism. I know, though I think `size` can also be 0 as a DNDarray might also contain 0 elements/be empty. But if that case is already handled above I guess that doesn't matter.